스피커, 리스너의 히스토리카드로 채널 입장 가능

happydhKim opened the pull request:
# [Epic#8, Epic#10] 스피커, 리스너의 히스토리카드로 채널 입장 가능

## 해당 이슈 📎

-[Epic#8, Epic#10] 스피커, 리스너의 히스토리카드로 채널 입장 가능 (#44)

## 변경 사항 🛠

```javascript
const getChannel = async (_, { channelId }, { user }) => {
  try {
    const channel = await Channels.findOne({ channelId });
    const { userId } = user;
    const { masterId } = channel;
    const master = await Users.findOne({ userId: channel.masterId });
    const status = channel ? 'ok' : 'not_exist';
    const isMaster = !!channel && !!user && channel.masterId === user.userId;
    const findHistory = await Histories.find({ userId, channelId });
    const firstVisitedChannel = () => !isMaster && findHistory.length === 0;
    if (firstVisitedChannel()) {
      const newHistory = new Histories({
        userId,
        masterId,
        channelId,
        updatedAt,
      });
      await newHistory.save();
    }
    if (!channel) return { status, isMaster };

    const payload = await channel.toPayload({ master });

    return {
      status,
      isMaster,
      channel: payload,
    };
  } catch (err) {
    throw new ApolloError(err.message);
  }
};
```

- 채널의 정보를 가져올 때, 처음 방문인 채널인 경우 히스토리에 추가하도록 했습니다. 하지만 함수 단위가 커져서 고민입니다. 🤔 어떻게 생각하시는지 이 쪽을 중점적으로 리뷰를 부탁드립니다.
- 이제 라우팅 효과가 정상적으로 작동 됩니다.
- 히스토리가 정상적으로 쌓입니다.
- 카드 간격이 조금 더 자연스러워 졌습니다.

## 테스트 ✨

>없음

## 리뷰어 참고 사항 🙋‍♀️

> 없음

happydhKim committed:
[ FEATURE ] 프로필 메뉴를 누르면 마이페이지로 이동되도록 구현 (#44)

Why
- 마이페이지로 이동할 수 있어야 한다.

What
- 헤더의 프로필 메뉴로 mypage로 이동 할 수 있도록 구현.

happydhKim committed:
[ FEATURE ] 히스토리 카드 클릭시 해당 채널로 이동 구현 (#44)

Why
- 히스토리 카드를 눌렀을 때, 해당 채널로 이동해야 한다.

What
- 해당 채널 ID 값으로 카드 클릭 시 이동 되도록 구현.

happydhKim committed:
[ FEATURE ] 리스너가 채널 입장 시 채널 기록 추가 (#44)

Why
- 리스너가 채널에 입장하면 히스토리 기록이 남아야 한다.

What
- 채널 정보를 가져오는 resolvers의 getChannel 메서드에서 마스터를 먼저 확인.
- 마스터가 본인이 아니고, 히스토리 기능이 없다면 히스토리에 추가.

happydhKim committed:
[ REFACTOR ] 마이페이지 리펙토링 (#44)

Why
- 채널이 중복되는 현상 수정이 필요하다.
- 히스토리 카드 간격 및 스타일 리펙토링이 필요하다.

What
- 헤더에서 프로필을 눌렀을 때, 안닫히던 현상 수정.
- 히스토리 카드 간격 및 스타일 일관성있게 수정.
- 채널이 중복되어 입력되는 현상 수정. ( 채널 체크시 함께 적용 )

aereeeee commented on server-api/src/apollo/resolvers/channels.js:
```diff
@@ -60,10 +61,22 @@ const createChannel = async (_, {
+        updatedAt,
+      });
+      await newHistory.save();
+    }
```
getChannel은 채널정보를 쿼리하는 resolver인데 여기에 history 데이터를 업데이트하는 로직이 포함되는것이 어색한것 같습니다. 또 history는 채널과 다른 도메인으로 타입이 따로 정의되어 있는 만큼 resolver도 따로 작성하는게 맞다고 생각합니다:)

happydhKim replied:
넵 감사합니다. 일단 나누어서 작업을 진행했습니당!!🐤

aereeeee commented on web/src/components/history/UserHistory/UserHistory.jsx:
```diff
@@ -17,6 +17,7 @@ const UserHistory = (props) => {
+      channelId={historyInfo.channel.channelId}
       channelStatus={historyInfo.channel.channelStatus}
       updatedAt={historyInfo.updatedAt}
       channelName={historyInfo.channel.channelName}
```
distructuring을 사용하면 더 깔끔할것 같아요~!

happydhKim replied:
넹 🥺 수정할게욥!!

aereeeee commented on web/src/components/history/UserHistoryCard/UserHistoryCard.jsx:
```diff
@@ -1,53 +1,56 @@
-  const color = channelStatus === 'on' ? 'primary' : 'default';
+  const presentationStatus = channelStatus === 'on'
+    ? { label: 'presentation-on', color: 'primary' }
+    : { label: 'presentation-off', color: 'default' };
```
status 버튼 구현할때 생각만하고 처리안해놓은 부분인데 이렇게 처리하셨군요! 좋은 것 같아요!bb

happydhKim replied:
헤헤헤 감사해용😁

happydhKim committed:
[ REFACTOR ] 리뷰 받은 사항 도메인별 resolver 나누는 부분 수정 (#44)

Why
- 리뷰 받은 사항 중 도메인별 resolver를 나눠야 하는 상황 발생.

What
- 도메인을 나누기 위해 resolvers/histories.js 에 addHistory를 추가.
- history mutation을 추가.
- 기존에 channel에 있던 history 관련 리팩토링.

happydhKim committed:
[ REFACTOR ] 히스토리카드 파라미터 리팩토링 (#44)

Why
- 리뷰에서 파라미터 리팩토링 요구를 처리해야 함.

What
- 기존 방식에서 destructuring으로 변경.

happydhKim committed:
[ UTIL ] 날짜를 표현해주는 util 구현 (#44)

Why
- 히스토리에 날짜를 보기 쉽게 표현해 줘야 한다.

What
- util에 dateParser함수 구현.
- 히스토리 카드에 연동.
